Graphs/README.md

```diff
@@ -0,0 +1,188 @@
+
+Additionally, in order to fold a graph, we need some way to indicate which node we care about the result for. This can be managed externally, but we can also have what's called a "rooted graph", where one or more nodes is blessed as a "root" node, giving us some way to extract the values we care about. We can model these roots fairly abstractly, as `Traversable r => r Int`, where `r` is often `Identity` (for a singly-rooted graph) or `[]` (for a multi-rooted graph), and the `Int`s point to the index of the root nodes.
+
+```haskell
+-- | A graph with vertices of type `v ()` and some structure `r` of roots.
+newtype Graph v = Graph 
+  { -- | Each entry in the list is a vertex in the graph. The `Int`s represent
+    --   edges from the current vertex to the vertex at the index indicated by
+	--   the `Int`.
+    vertices :: [v Int],
+  }
+  deriving (Generic)
+
+data MultiplyRootedGraph r v = MultiplyRootedGraph
+  { graph :: Graph v,
+    -- | The roots of the graph are some structure (often `Identity` or `[]`)
+	--   indicating vertices that are "blessed" as some starting point of the
+	--   graph. If the graph was constructed from `eliminateCommonSubtrees`,
+	--   each root corresponds to the root of one of the trees passed to that
+	--   function.
+    roots :: r Int
+  }
+  deriving (Generic)
+
+type RootedGraph v = MultiplyRootedGraph Identity v
+```
+
+We can then take advantage of our usual `f a -> a` algebras by traversing the list of vertices, building a parallel list with the `a` in place of the `f Int`, and looking up the indices of the children as we go. If our graph is rooted, we finish by looking up the `a` for each root index.
+
+## Implementation
+
+```haskell
+-- | A graph with vertices of type `v ()` and some structure `r` of roots.
+data Graph r v
+  = Graph
+      { -- | Each entry in the list is a vertex in the graph. The `Int`s
+	    --   represent edges from the current vertex to the vertex at the index
+		--   indicated by the `Int`.
+        graphVertices :: [v Int],
+        -- | The roots of the graph are some structure (often `Identity` or
+		--  `[]`) indicating vertices that are "blessed" as some starting point
+		--   of the graph. If the graph was constructed from
+		--  `eliminateCommonSubtrees`, each root corresponds to the root of one
+		--   of the trees passed to that function.
+        graphRoots :: r Int
+      }
+  deriving (Generic)
+
+-- | Used internally to hold information we need while constructing a graph from
+--   trees.
+data GraphState f
```

Comment:
I think we could spin more story around `GraphState`, eliminateCommonSubtree`, and `findOrAddNode`. So again, what's the motivation for each step. What are the smaller sub-problems we're trying to solve? <--- this is one of our general motifs :smile: